### ISSUE TITLE
Inconsistent type inference with closure and ensure

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#6037]
```cr
begin
  foo = "foo"
  -> do
    typeof(foo) # => String
    foo.upcase # undefined method 'upcase' for Nil (compile-time type is (String | Nil))
  end.call
ensure
  if foo
  end
end
```
[carc.in](https://carc.in/#/r/3z6l/edit)

Calling `foo.upcase` raises because compile-time type is `String | Nil` but, (with that line commented out), `typeof` correctly reports `String`. `foo` can't really be `Nil` in the closure, because the value is never changed. Without the `ensure` block it compiles.

It seems that `if foo` in the `ensure` block triggers some error in the type inference, even if it does pretty much nothing, at least not change the variable or it's type.

**Comment by asterite** [comment]
Yes, closures are a mess in the language. The best advise is to workaround all of that, somehow.
